test/python/circuit/library/test_phase_and_bitflip_oracles.py

```diff
@@ -0,0 +1,234 @@
+        ("x & x", "0", False),
+        ("(x0 & x1 | ~x2) ^ x4", "0110", False),
+        ("xx & xxx | ( ~z ^ zz)", "0111", True),
+    )
+    @unpack
+    def test_evaluate_bitstring(self, expression, input_bitstring, expected):
+        """PhaseOracle(...).evaluate_bitstring"""
+        oracle = PhaseOracle(expression)
+        result = oracle.evaluate_bitstring(input_bitstring)
+        self.assertEqual(result, expected)
+
+    @data(
+        ("x | x", "01"),
+        ("~x", "10"),
+        ("x & y", "0001"),
+        ("x & ~y", "0100"),
+        ("(x0 & x1 | ~x2) ^ x4", "1111000100001110"),
+        ("x & y ^ ( ~z1 | z2)", "1110000111101110"),
+    )
+    @unpack
+    def test_statevector(self, expression, truth_table):
+        """Circuit generation"""
+        oracle = PhaseOracle(expression)
+        num_qubits = oracle.num_qubits
+        circuit = QuantumCircuit(num_qubits)
+        circuit.h(range(num_qubits))
+        circuit.compose(oracle, inplace=True)
+        statevector = Statevector.from_instruction(circuit)
+
+        valid_state = -1 / sqrt(2**num_qubits)
+        invalid_state = 1 / sqrt(2**num_qubits)
+
+        states = list(range(2**num_qubits))
+        good_states = [i for i in range(len(states)) if truth_table[i] == "1"]
+        expected_valid = [state in good_states for state in states]
+        result_valid = [isclose(statevector.data[state], valid_state) for state in states]
+
+        expected_invalid = [state not in good_states for state in states]
+        result_invalid = [isclose(statevector.data[state], invalid_state) for state in states]
+        self.assertListEqual(expected_valid, result_valid)
+        self.assertListEqual(expected_invalid, result_invalid)
+
+    @data(
+        ("((A & C) | (B & D)) & ~(C & D)", None, [3, 7, 12, 13]),
+        ("((A & C) | (B & D)) & ~(C & D)", ["A", "B", "C", "D"], [5, 7, 10, 11]),
+    )
+    @unpack
+    def test_variable_order(self, expression, var_order, good_states):
+        """Circuit generation"""
+        oracle = PhaseOracle(expression, var_order=var_order)
+        num_qubits = oracle.num_qubits
+        circuit = QuantumCircuit(num_qubits)
+        circuit.h(range(num_qubits))
+        circuit.compose(oracle, inplace=True)
+        statevector = Statevector.from_instruction(circuit)
+
+        valid_state = -1 / sqrt(2**num_qubits)
+        invalid_state = 1 / sqrt(2**num_qubits)
+
+        states = list(range(2**num_qubits))
+        expected_valid = [state in good_states for state in states]
+        result_valid = [isclose(statevector.data[state], valid_state) for state in states]
+
+        expected_invalid = [state not in good_states for state in states]
+        result_invalid = [isclose(statevector.data[state], invalid_state) for state in states]
+        self.assertListEqual(expected_valid, result_valid)
+        self.assertListEqual(expected_invalid, result_invalid)
+
+
+@ddt
+class TestPhaseOracleGate(QiskitTestCase):
```

Comment:
Could you merge this test with the one for the circuit? They seem to be the same 🙂 You could e.g. do
```python
@data( ... )
@unpack
def test_statevector(self, expression, truth_table):
    for use_gate in [True, False]:
        oracle = PhaseOracleGate(expression) if use_gate else PhaseOracle(expression)
        # ... rest as before ...
        
        with self.subTest(use_gate=use_gate):
            self.assertListEqual( ... )
```

Reply:
I actually prefer to avoid this since we are deprecating `PhaseOracle` and I want to keep is separate from `PhaseOracleGate`.

Reply:
We're only going to remove it for 3.0 though, so it would be nice to have them together since bugfixes and new checks would likely need to cover both. But we can also keep them separate and copy these, if they occur in the future, in case you prefer 🙂

Reply:
I'm convinced. Done in 454ecc5ee